flocker/node/agents/test/test_blockdevice.py

```diff
@@ -698,6 +752,186 @@ def add_application_with_volume(node_state):
+
+
+def _infer_volumes_for_test(node_uuid, node_hostname, compute_instance_id,
+                            cluster_state):
+    """
+    Infer the current ``BlockDeviceVolume`` instances in the cluster. This is
+    for use by tests that create cluster_state by hand instead of using the API
+    to actually construct the cluster.  The ``BlockDeviceVolume`` instances
+    constructed with this function do not have test blockdevice_ids that can be
+    used with the ``IBlockDeviceAPI``.
+
+    If you are using the ``IBlockDeviceAPI`` to set up the cluster_state, just
+    use api.list_volumes() instead of calling this function.
+
+    :param UUID node_uuid: The uuid of the local node.
+
+    :param unicode node_hostname: The hostname of the local node.
+
+    :param unicode compute_instance_id: The value for the attached_to field of
+        the ``BlockDeviceVolume`` instances that are created for manifestions
+        on the local node.
+
+    :param DeploymentState cluster_state: The rest of the cluster_state. This
+        is inspected to determine the volumes that are not attached to any node
+        (nonmanifest_datasets) and the volumes that are attached to remote
+        nodes.
+
+    :returns: A list of ``BlockDeviceVolume`` instances for all the volumes in
+        cluster_state that is only suitable for use in tests that do not use
+        the ``IBlockDeviceAPI`` to set up the cluster_state.
+    """
+    node_state = cluster_state.get_node(node_uuid,
+                                        hostname=node_hostname)
+    local_volumes = [
+        _create_test_blockdevice_volume_for_dataset_id(
+            dataset_id, attached_to=compute_instance_id)
+        for dataset_id in (node_state.manifestations or {}).keys()]
+
+    nonmanifest_datasets = cluster_state.nonmanifest_datasets
+    unattached_volumes = [
+        _create_test_blockdevice_volume_for_dataset_id(dataset_id)
+        for dataset_id in (nonmanifest_datasets or {}).keys()]
+
+    volumes = local_volumes + unattached_volumes
+
+    for remote_node_state in cluster_state.nodes:
+        if remote_node_state == node_state:
+            # Skip this node for remote processing if it is the local node.
+            continue
+        # Construct a fake instance id for the attached_to field of the
+        # ``BlockDeviceVolume`` instances that are connected to remote nodes.
+        # This is an opaque identifier, and just needs to be different than
+        # compute_instance_id for the sake of testing this node.
+        fake_instance_id = u"remote_node_{}".format(
+            unicode(remote_node_state.uuid))
+        remote_volumes = [
+            _create_test_blockdevice_volume_for_dataset_id(
+                dataset_id, attached_to=fake_instance_id)
+            for dataset_id in (remote_node_state.manifestations or {}).keys()]
+
+        volumes.extend(remote_volumes)
+    return volumes
+
+
+# Constants for determining volumes in calculate_changes tests. See
```

Comment:
I believe you need to have these as class attributes (http://twistedmatrix.com/documents/current/core/howto/constants.html)